sdk/core/Azure.Core/src/Serialization/AzureObjectSerializer.cs

```diff
@@ -0,0 +1,48 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.IO;
+using System.Threading.Tasks;
+
+
+namespace Azure.Core.Serialization
+{
+    /// <summary>
+    /// An abstraction from reading typed objects.
+    /// </summary>
+    public abstract class AzureObjectSerializer
+    {
+        /// <summary>
+        /// Convert the provided value to it's binary representation and write it to <see cref="System.IO.Stream"/>.
+        /// </summary>
+        /// <param name="stream">The <see cref="System.IO.Stream"/> to write to.</param>
+        /// <param name="value">The value to convert.</param>
+        /// <param name="inputType">The type of the <paramref name="value"/> to convert.</param>
+        public abstract void Serialize(Stream stream, object? value, Type inputType);
+
+        /// <summary>
+        /// Convert the provided value to it's binary representation and write it to <see cref="System.IO.Stream"/>.
+        /// </summary>
+        /// <param name="stream">The <see cref="System.IO.Stream"/> to write to.</param>
+        /// <param name="value">The value to convert.</param>
+        /// <param name="inputType">The type of the <paramref name="value"/> to convert.</param>
+        public abstract ValueTask SerializeAsync(Stream stream, object? value, Type inputType);
+
+        /// <summary>
+        /// Read the binary representation into a <paramref name="returnType"/>.
+        /// The Stream will be read to completion.
+        /// </summary>
+        /// <param name="stream">The <see cref="System.IO.Stream"/> to read from.</param>
+        /// <param name="returnType">The type of the object to convert to and return.</param>
+        public abstract object Deserialize(Stream stream, Type returnType);
+
+        /// <summary>
+        /// Read the binary representation into a <paramref name="returnType"/>.
+        /// The Stream will be read to completion.
+        /// </summary>
+        /// <param name="stream">The <see cref="System.IO.Stream"/> to read from.</param>
+        /// <param name="returnType">The type of the object to convert to and return.</param>
+        public abstract ValueTask<object> DeserializeAsync(Stream stream, Type returnType);
```

Comment:
Why return object? Won't this cause everyone to constantly cast? Is there any reason to avoid generics?

public abstract ValueTask<T> DeserializeAsync<T>(Stream stream);

Reply:
Generic is more modern for most common scenarios. Any reason for avoiding them?

Reply:
Yes, but generics are also "viral" requiring all method in the call chain to also be generic and limiting amount of scenarios that this abstraction can support.

Generic have one benefit over returning `object` and that's avoiding boxing but it's not a concern here because we expect this abstraction to be used mostly to operate on classes.

In addition this abstraction is not expected to be directly consumed by end-users of the Azure SDKs, it would always be wrapped into a generic client library method so ease-of-use is also not a concern.

You can always implement generic method using signature like this but doing an opposite is very hard.

Reply:
Can you provide an example where generics limit the scenarios, but object doesn't?

You are still requiring the caller to pass in the type, so I don't see the benefit. This just introduces the possibility of casting exceptions.

`Book book = (Book)await serializer.DeserializeAsync(new MemoryStream, typeof(Book));`
`Book book = await serializer.DeserializeAsync<Book>(new MemoryStream);`

Reply:
Let's say we want to deserialize the payload as part of the HttpPipeline policy.

Policies are not generic so there is no way to call `Deserialize<T>` but you can always pass `Type` instance into policy constructor and use it.

Reply:
Any reason they can't just do the following and have the implementation class have the generic?

`public class ValidateType<T> : HttpPipelinePolicy {}`

Reply:
You can do that, but now you need a generic ClientOptions<T> type to be able to pass the T down to policy construction.

This is what I mean by viral generics.

Reply:
Users still have to pass the type into all these classes. All this is doing is avoiding using the normal c# functionality built to handle this. By avoiding using generics it introduces painful run time exceptions, makes the code difficult to read, and limits the functionality users can do. 

Current model how can users use the generic object?
`new SomeClientOptions(typeof(Book))`
`Object object = await serializer.DeserializeAsync(memoryStream, type);`

With generics users can do validation and even modify the deserialized object.
`new SomeClientOptions<Book>()`
`Book book = await serializer.DeserializeAsync<Book>(memoryStream);`

Reply:
In Cosmos DB we allow the user to build their own custom serializer through their own implementation of `CosmosSerializer`. In this case, this type would be made public to let the user provide their own custom implementation. You mentioned that this abstraction is not expected to be consumed by end-users, so this case wouldn't be supported?

Reply:
@pakrym, would it make sense to add:
`public virtual T Deserialize<T>(Stream stream) => (T)Deserialize(stream, typeof(T));`

Reply:
@KrzysztofCwalina we could but what's the benefit? It's only saving 10 characters on the call site.

> You mentioned that this abstraction is not expected to be consumed by end-users, so this case wouldn't be supported?

The case is supported, what I meant is that end customers would never **call** Deserialize method on this type so they won't have to manually cast. In terms of implementing the serializer there is not much difference between generic or non generic option.

Reply:
> Current model how can users use the deserialized object?
> With generics users can do validation and even modify the deserialized object.

End users should never call members of this class. The intent is that other libraries would integrate with it and provide generic overloads of their own that customers would use.

Reply:
There still doesn't seem any benefit for using object and Type in the parameter. Generics and Type parameter in the CTOR are going to be applied to all the same classes. Using generics seems to have a lot of benefits over Object and Type in the constructor.

Reply:
There is a huge benefit of not requiring all methods in the call site to be generic. You can always implement the generic version using a non generic one but not the other way around and we would like to keep as much flexibility is possible with such a core abstraction.

What benefits of generic version do you see considering the type won't be used by customers directly?

Reply:
1. It avoids casting exceptions and other runtime issues that are extremely painful
2. It allows users to work on the type.
```
public class ValidateType<T> : HttpPipelinePolicy where T is Animal {}
Animal animal = await serializer.DeserializeAsync<T>(memoryStream);
this.Validate(animal)
```
3. Users can always use generic object like JObject if they don't want to use the generic.
4. It increases readability by making it clear that the class can only handle specific types. Passing the type in makes it ambiguous what that is being used for.

```
public class ValidateType<T> : HttpPipelinePolicy {}
new ValidateType<Animal>();
```
Vs.
```
public class ValidateType : HttpPipelinePolicy {}
new ValidateType(typeof(Animal));
```

Reply:
> It avoids casting exceptions and other runtime issues that are extremely painful

If you pattern is `(T)serializer.Deserialize(typeof(T))` there is no possibility for runtime cast exceptions.

> Users can always use generic object like JObject if they don't want to use the generic.

Sorry, Not sure I understand this point.

> 2.4.
You highlighted how both of the scenarios can be implemented using both models.

But how would you implement a `ValidateType` policy that's able to handle multiple types? It's simply impossible to do without reflection when only a generic method is available

```
public class ValidateTypes : HttpPipelinePolicy {}

ValidateType(Type[] types)
{
}

```